src/data-store.js

```diff
@@ -0,0 +1,68 @@
+// An in-memory object that implements getItem/setItem that we can use in place
+// of localStorage when localStorage is not available.
+const createInMemoryStorage = () => {
+  const data = {};
+  return {
+    setItem: (key, value) => {
+      data[key] = value;
+    },
+    getItem: key => {
+      if (data.hasOwnProperty(key)) return data[key];
+      return null;
+    },
+  };
+};
+
+// Tests if localStorage is available. There are a number of interesting cases
+// to consider across browsers. Rather than just checking the property it's
+// important to actually test setItem due to some behavior of Safari in previous
+// versions.
+//
+// Reference:
+// https://michalzalecki.com/why-using-localStorage-directly-is-a-bad-idea/
+const isLocalStorageSupported = () => {
+  try {
+    const testKey = `enlearn.localStorageSupportTest`;
+    window.localStorage.setItem(testKey, "");
+    window.localStorage.removeItem(testKey);
+    return true;
+  } catch (err) {
+    return false;
+  }
+};
+
+// A wrapper around an implementation of Storage to give us a Promise based
+// interface and key scoping. The read/write methods are the interface on which
+// the Enlearn Client depends.
+export const createDataStore = appId => {
+  const fullKey = key => `enlearn.${appId}.${key}`;
```

Comment:
Would you foresee any issues if the `appId` or `key` strings contained a "."? It looks like this is just to avoid key collision.

Reply:
Shouldn't be a problem. Like you mentioned this was just to avoid key collision if two apps have the same origin and therefore share local storage. Using the appId was just to help us differentiate the two apps.

Reply:
I will add I had considered adding a second parameter for student ID so that we could also store student-specific data as part of the key, that way we'd be setup for games that want to support different students (or just to deal with weird issues of the student ID getting reset). But that seemed not necessary at this time.